csrc/device_lower/analysis/sync_information.cpp

```diff
@@ -662,16 +679,67 @@ SyncMap::SyncMap(Fusion* fusion) {
             continue;
           }
 
+          // Check if the IdModel-based analysis matches with the
+          // existing analysis
+          bool requires_sync = true;
+
+          IterDomain* p_loop = nullptr;
+          IterDomain* c_loop = nullptr;
+
+          if (GpuLower::current()->hasIdModel()) {
+            if (producer_ptype == consumer_ptype) {
+              // If p_id and c_id are mapped in BestEffortReplay with
+              // no broadcast forwarding, they should not require any
+              // synchronization.
+              if (auto it = p2c_map_no_forwarding.find(p_id);
+                  it != p2c_map_no_forwarding.end() && it->second == c_id) {
+                requires_sync = false;
```

Comment:
Why do we need this branch? Isn't the second branch itself sufficient?

Reply:
That was my initial thought, but it's actually not enough since the AlmostExact graph doesn't map broadcast and non-broadcast domains. For example, just a simple fusion like this would fail:

```
// t0: [i0]
// t1: [b1]
t2 = t0 + t1
// t2: [i2]
```

Suppose nothing is inlined, so there's no promotion. Since `b1` and `i2` are not mapped, this would result in requiring a synchronization, which isn't the case. `t2` and `t1` use different indices but no sync is needed.

That's why I also mentioned about the broadcast graph. If it were used instead of the AlmostExact graph, that would be sufficient as well.

I think in this case `BestEffortReplay` would be good enough.

Reply:
If t1 is on smem, and both b1 and i2 are parallelized on TIDx, then we wouldn't need sync? For this case will 1 thread write smem, or all threads write it? Whichever it is, don't we need to wait until all writes are complete before reading?

Reply:
In that case, TIDx would be marked as a redundant parallel type and only one thread should write to smem. You're right a sync would be required, which should be taken care at lines around 580. The way how it's handled doesn't seem ideal to me, but I'm trying to make an incremental improvement.